File: 06_breadth_search/python/01_bredth_search.py
'''
@Description: In User Settings Edit
@Author: your name
@Date: 2019-08-26 14:34:01
@LastEditTime: 2019-08-26 14:59:04
@LastEditors: Please set LastEditors
'''
from collections import deque


graph = {}
graph["you"] = ["alice", "bob", "claire"]
graph["bob"] = ["anuj", "peggy"]
graph["alice"] = ["peggy"]
graph["claire"] = ["thom", "jonny"]
graph["anuj"] = []
graph["peggy"] = []
graph["thom"] = []
graph["jonny"] = []

def person_is_seller(name):
    return name[-1] == 'm'

def search(name):
    search_queue = deque()
    search_queue += graph[name]

    searched = []

    while len(search_queue) :
        person = search_queue.popleft()
        if person not in searched :
            if  person_is_seller(person):
                print(person)
                return True
            
            search_queue += graph[person]
            searched.append(person)
    
    return False

search('you')

File: 06_breadth_search/ts/01_bredth_search.ts
/*
 * @Description: In User Settings Edit
 * @Author: your name
 * @Date: 2019-08-26 14:25:23
 * @LastEditTime: 2019-08-26 14:31:02
 * @LastEditors: Please set LastEditors
 */
const graph = {};
graph["you"] = ["alice", "bob", "claire"];
graph["bob"] = ["anuj", "peggy"];
graph["alice"] = ["peggy"];
graph["claire"] = ["thom", "jonny"];
graph["anuj"] = [];
graph["peggy"] = [];
graph["thom"] = [];
graph["jonny"] = [];

const person_is_seller = (name:string) => name[name.length-1] === 'm'


const search = (name:string) => {
    let search_queue:any[] = [];

    search_queue = search_queue.concat(graph[name])
    let searched:any[]=[];

    while(search_queue.length){
        let person = search_queue.shift();

        if(searched.indexOf(person) === -1){
            if(person_is_seller(person)){
                console.log(person)
                return true;
            }
            search_queue = search_queue.concat(graph[person])
            searched.push(person)
        }
    }

    return false;
}

search('you')